Update documentation

ttuffin opened the pull request:
Update all documentation as required for #156.

TODO:
- [x] `contributing.rst` @ttuffin 
- [x] Update `README.md` @ttuffin 
- [x] remove `decorators.rst` and integrate into `plugins`
- [x] merge `installation.rst`, `deploy.rst` and `usage.rst` into single deployment doc @ttuffin 
- [x] update `dev-install.rst` (@ttuffin is working on this). I have already updated `testing.rst`, but perhaps we should merge these two into a single `development` doc? ttuffin: has been merged into `contributing.rst`
- [x] Update `settings.rst` (I'll take care of this, but I'm not sure we even need it. The `Settings` class is pretty self-explanatory, so perhaps we could just add a link to the source code in the `deploy` section.
- [x] Update `use_cases.rst`, but I'm not a fan of this document at all to be honest. Again, the `ExamplePlugin` is quite self-explanatory, so if we just link to that source code in one of the other docs we should be good.
- [ ] Create `migration.rst` to document 1.x -> 2.x process. @unode  
- [x] Integrate `deploy.rst` into `getting_started.rst` and delete the old file.

ttuffin commented:
Started with plugins documentation, still need to update the rest. Feel free to update/change as you see fit.

attzonko commented on docs/dev_install.rst:
```diff
@@ -1,20 +1,38 @@
+        $ podman-compose -f tests/integration_tests/docker-compose.yml up -d
+
+#. Edit mmpy_bot/settings.py and configure as below. If testing webhook
+   server functionality, ensure **WEBHOOK_HOST_ENABLED** is to `True`.
```
missing word "set"

attzonko commented on docs/dev_install.rst:
```diff
@@ -1,20 +1,38 @@
 
+        $ git clone https://github.com/attzonko/mmpy_bot.git
+        $ cd mmpy_bot
+        $ pip install -r requirements.txt
+        $ pip install -r dev-requirements.txt
```
Why don't we provide the same command used by the workflow `pip install -e  .[dev]`

attzonko commented on docs/dev_install.rst:
```diff
@@ -1,20 +1,38 @@
 ============================
 
-.. code-block:: bash
+#. Clone repo and install requirements:
```
Do we want to add documentation on using `venv` module, to isolate the development environment from their default python environment?

attzonko commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
+                     MyPlugin()]
+        ):
 
+   If everything went as planned you can now start your bot, send the message "wake up" and expect to appropriate reply.
```
should be "expect `the` appropriate reply."

attzonko commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
+            """Showcases a function with restricted access."""
+            self.driver.reply_to(message, "Access allowed!")
+
+Click support
```
We should write a quick blurb here explaining the benefit of using Click?

ttuffin replied:
Agreed, but to be honest, I don't even know myself what it's for :smile:. @jneeven can you write a quick couple of lines about it?

jneeven replied:
Click is amazing! I'll write something tomorrow :+1:

jneeven commented on docs/dev_install.rst:
```diff
@@ -1,20 +1,38 @@
+
+        $ podman-compose -f tests/integration_tests/docker-compose.yml up -d
+
+#. Edit mmpy_bot/settings.py and configure as below. If testing webhook
```
The idea is to *not* edit settings.py, but just to create a  `Settings` object where you specify the settings you want to change (e.g. like we do for the integration tests). But this will likely need an update after I've implemented the environment variables, so we can leave it like this for now.

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
+-------------------------
 
-A chat bot is meaningless unless you can extend/customize it to fit your own use cases.
+#. First you need to create a Python file in the `./plugins` directory. For this example we will create the file ./plugins/my_plugin.py and
```
Since a custom bot does not require checking out the source code (I'd even recommend against it), there won't be a plugins directory. Users can just subclass `Plugin` wherever they want and pass it to their `Bot` constructor (as with the settings)

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
-- A function decorated with ``respond_to`` is called when a message matching the pattern is sent to the bot (direct message or @botname in a channel/group chat)
-- A function decorated with ``listen_to`` is called when a message matching the pattern is sent on a channel/group chat (not directly sent to the bot)
-- A function decorated with ``at_start`` is called as soon as the plugin is initialized (when the bot starts)
+        from mmpy_bot.plugins.base import Plugin, listen_to
```
```suggestion
        from mmpy_bot import Plugin, listen_to
```

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
-- A function decorated with ``listen_to`` is called when a message matching the pattern is sent on a channel/group chat (not directly sent to the bot)
-- A function decorated with ``at_start`` is called as soon as the plugin is initialized (when the bot starts)
+        from mmpy_bot.plugins.base import Plugin, listen_to
+        from mmpy_bot.wrappers import Message
```
```suggestion
        from mmpy_bot import Message
```

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
+    .. code-block:: python
 
-    @listen_to('Can someone help me?')
-    def help_me(message):
-        # Message is replied to the sender (prefixed with @user)
-        message.reply('Yes, I can!')
+        from mmpy_bot.plugins.base import Plugin
+        from mmpy_bot.plugins.example import ExamplePlugin
+        from mmpy_bot.plugins.webhook_example import WebHookExample
+        from mmpy_bot.plugins.my_plugin import MyPlugin
 
-        # Message is sent on the channel
-        # message.send('I can help everybody!')
+        __all__ = ["Plugin", "ExamplePlugin", "WebHookExample", "MyPlugin"]
 
-    @at_start
-    def hello(client):
-        # Note that contrary to respond_to and listen_to @at_start
-        # receives a client object and not a message object
-        team = client.api.get_team_by_name("TESTTEAM")
-        channel = client.api.get_channel_by_name("bot_test", team["id"])
-        client.channel_msg(channel["id"], "Hello, feels good to be alive!!")
```
This is not necessary unless they are contributing

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
+#. Last but not least, you will need to ensure your plugin is started by the bot. In this example we will update the default `bot.py` file
+   to import `MyPlugin`:
 
-To extract params from the message, you can use regular expression:
+    .. code-block:: python
 
-.. code-block:: python
+        from mmpy_bot.plugins import ExamplePlugin, Plugin, WebHookExample, MyPlugin
 
-    from mmpy_bot.bot import respond_to
+        def __init__(self, settings=Settings(), plugins=[ExamplePlugin(), WebHookExample(),
+                     MyPlugin()]
```
As mentioned above, they don't need to modify any files other than their main file. They can just have a file like this:

```python
from mmpy_bot import Bot, Settings
from my_plugin import MyPlugin

bot = Bot(settings=Settings(BOT_TOKEN="my_token"), plugins=[MyPlugin()])
bot.run()
```

ttuffin replied:
Thanks, that makes a lot more sense!

jneeven replied:
I've created an `entrypoint.py` in #162 that we can refer to in the docs

jneeven commented on docs/plugins.rst:
```diff
@@ -1,260 +1,133 @@
 
-mmpy_bot integrates `schedule 
-<https://github.com/dbader/schedule/>`_ to provide in-process job scheduling.
-
-With `schedule 
-<https://github.com/dbader/schedule/>`_, we can put periodic jobs into waiting queue like this:
-
-.. code-block:: python
-
-    import re
-    from datetime import datetime
-    from mmpy_bot.bot import respond_to
-    from mmpy_bot.scheduler import schedule
-
-
-    @respond_to('reply \"(.*)\" every (.*) seconds', re.IGNORECASE)
-    def reply_every_seconds(message, content, seconds):
-        schedule.every(int(seconds)).seconds.do(message.reply, content)
-
-
-    @respond_to('cancel jobs', re.IGNORECASE)
-    def cancel_jobs(message):
-        schedule.clear()
-        message.reply('all jobs canceled.')
-
-The `schedule 
-<https://github.com/dbader/schedule/>`_ itself provide human-readable APIs to schedule jobs. Check out `schedule.readthedocs.io <https://schedule.readthedocs.io/>`_ for more usage examples.
-
-`schedule 
-<https://github.com/dbader/schedule/>`_ is designed for periodic jobs.
-In order to support one-time-only jobs, mmpy_bot has a monkey-patching on integrated 
-`schedule 
-<https://github.com/dbader/schedule/>`_ package.
-
-We can schedule a one-time-only job by `schedule.once` method.
-You should notice that this method takes a datetime object, which is different from `schedule.every` methods.
-
-The following code example uses `schedule.once` to schedule a job.
-This job will be trigger at `t_time`.
-
```
All of this is still relevant, I've only added the `subprocess` tag

jneeven reviewed (commented):
Awesome, thanks  for getting this started!

ttuffin commented:
Great feedback :+1: I will update with the proposed changes later today.

jneeven commented:
I've updated two of the docs and added a to-do list for the remaining ones in https://github.com/attzonko/mmpy_bot/pull/161#issue-592115248. Let me know what you think about those proposals!

ttuffin commented:
> I've updated two of the docs and added a to-do list for the remaining ones in [#161 (comment)](https://github.com/attzonko/mmpy_bot/pull/161#issue-592115248). Let me know what you think about those proposals!

Agree with you in regards to dev-install.rst and testing.rst. I am updating contributing.rst to combine both of these topics, since they all fall under the contributor umbrella.

ttuffin commented on README.md:
```diff
@@ -64,175 +64,33 @@ Note: **Enable Bot Account Creation** must be enabled under System Console
 
 For use all API(V4.0.0), you need add bot user to system admin group to avoid 403 error.
```
Is this still valid? It's from the legacy README.md but I don't think this is the case anymore?

jneeven replied:
The description is a bit cryptic, but the bot does indeed need system admin permissions to use the entire API, e.g. ephemeral replies.

ttuffin replied:
Cool, I've updated the doc to clarify this a little better.

unode commented:
Hi everyone, I'm trying to update my existing code and I'm finding a few cases that might benefit from a "migrating from 1.0 to 2.0" kind of document. A few questions so far:

1. Is the "Django-like" `settings.py` completely gone? If I had extra stuff in my `settings.py` do I need to subclass the `Settings` dataclass? 
2. Do I understand correctly that the `click` interface is meant to be used via mattermost and not for command-line invocations?
3. There's now only `listen_to`. I see a `needs_mention=` argument that seems to make it behave like `respond_to`. Is there anything like [`at_start`](https://mmpy-bot.readthedocs.io/en/latest/plugins.html?highlight=at_start#id1) hiding somewhere?

jneeven commented:
> a "migrating from 1.0 to 2.0" kind of document.

Very good idea!

> If I had extra stuff in my settings.py do I need to subclass the Settings dataclass?

This is indeed the way to do it. I'm not very familiar with Django, have I accidentally prevented any valid use-cases now?

> Do I understand correctly that the click interface is meant to be used via mattermost and not for command-line invocations?

Correct again ;) I found this much more robust and straightforward than messing with regular expressions if you want to execute commands that way. Of course, if you do want to use regular expressions and / or take a more "natural language" kind of approach, that's still supported.

> Is there anything like at_start hiding somewhere?

Yes, [there is indeed](https://github.com/attzonko/mmpy_bot/blob/master/mmpy_bot/plugins/base.py#L60). And an `on_stop` as well, should you need it.

If any more questions arise, please keep asking them! A "FAQ" like this is probably at least as valuable as any documentation updates will be :+1:

unode commented:
> This is indeed the way to do it. I'm not very familiar with Django, have I accidentally prevented any valid use-cases now?

I think not but I'll confirm this once I get 2.0 running locally with our 1.3 plugins. We had some secrets we were importing from `settings.py` and will likely need a different approach now.

ttuffin commented:
@unode that's great if you can document the migration from 1.x -> 2.x :+1: 

I'll add it to the to-do list.

ttuffin commented:
I think `installation.rst`, `deploy.rst` and `usage.rst` can be removed/merged in favour of one deployment document.

- `deploy.rst` is specific to Ubuntu (apt), and we should try to keep instructions distribution agnostic if possible.
- The current `installation.rst` is short and can be included in another document without making it overblown.

In essence, a single document that:
1. Details installation procedure via Git repo, pip, and containers.
2. How to run the bot and define settings and plugins through an `entrypoint` file.
3. A link to `plugins.rst` for writing custom plugins.

attzonko commented:
> I think `installation.rst`, `deploy.rst` and `usage.rst` can be removed/merged in favour of one deployment document.
> 
> * `deploy.rst` is specific to Ubuntu (apt), and we should try to keep instructions distribution agnostic if possible.
> * The current `installation.rst` is short and can be included in another document without making it overblown.
> 
> In essence, a single document that:
> 
> 1. Details installation procedure via Git repo, pip, and containers.
> 2. How to run the bot and define settings and plugins through an `entrypoint` file.
> 3. A link to `plugins.rst` for writing custom plugins.

Sounds good.

ttuffin commented on docs/getting_started.rst:
```diff
@@ -58,7 +58,7 @@ We recommend creating an entrypoint file for executing the bot, which will look
 
 You can then set the executable bit on this file and run the bot straight from the command line.
-For more information on configuring bot settings and plugins, please see the respective :doc:`settings` :doc:`plugins` pages.
+For more information on configuring bot settings and plugins, please see the respective :ref:`settings <settings>` and :ref:`plugins <plugins>` pages.
```
Not having a fun time with rst :disappointed:. Whats the syntax to link to an internal page?

ttuffin replied:
I realised this was just Github failing to render the file properly. It works in readthedocs.

jneeven commented on docs/getting_started.rst:
```diff
@@ -91,3 +91,35 @@ To check your installed version of `mmpy_bot`, simply open a Python interpreter
         print(mmpy_bot.__version__)
+
+
+Integration with Django
+-----------------------
+
+You can create a ``django`` command as follows:
+
+    from django.core.management.base import BaseCommand
+
+    from mmpy_bot import Bot, Settings
+
+
+    class Command(BaseCommand):
+
+        def handle(self, **options):
+            b = Bot(settings=Settings())
+            b.run()
+
+
+Modify ``manage.py``::
+
+    #!/usr/bin/env python
+
+    import os
+    import sys
+
+
+    if __name__ == "__main__":
+        os.environ.setdefault("DJANGO_SETTINGS_MODULE", "project.settings")
+        from django.core.management import execute_from_command_line
+
+        execute_from_command_line(sys.argv)
```
Can someone have a look at this? I have zero Django experience and this example probably doesn't work, but I'm not sure how Django works in the first place.

attzonko replied:
I look at this but don't really know what I am looking at 😸 Honestly I am not sure we need to include the Django integration section. What is the actual use case for that?

jneeven replied:
Yeah fair point, I wouldn't know either... I'll remove it altogether 👍🏼

ttuffin commented:
Almost there :+1: 
In regards to `deploy.rst`, the new doc (`getting-started.rst`) already covers installation via pip, repo and container. Using `supervisor` to run the bot seems like a corner-case to me and not necessary to include in docs.

ttuffin reviewed (commented):
@jneeven @unode feel free to add your e-mails if you wish.

attzonko commented on AUTHORS.rst:
```diff
@@ -2,20 +2,21 @@
-Development Lead
-----------------
-
+Development Leads
```
I don't like the `Development Lead` bucket at all. I kept it from the previous author. I would be more than happy if we just have an `Active Contributors`  and `Past Contributors`

ttuffin replied:
Done!

attzonko commented:
By the way thanks for all the hard work on updating the docs. Looks like the TODO list is done, minus the migration docs. I am not sure we need to wait for those before merging. What do you all think?

attzonko commented on README.md:
```diff
@@ -62,177 +62,37 @@ Note: **Enable Bot Account Creation** must be enabled under System Console
+        MATTERMOST_PORT = 443,
+        BOT_TOKEN = "a69155mvlsobcnqpfdceqihaa",
+        BOT_TEAM = "test",
+        SSL_VERIFY = 443,
```
SSL_VERIFY = 443?

ttuffin replied:
Oops, good catch. Corrected!

attzonko commented on docs/testing.rst:
```diff
@@ -4,118 +4,76 @@
 Testing
```
The info here seems to be duplicated from contributing.rst do we really need it in two places?

ttuffin replied:
Yep you are correct. That was one of the files on the list to remove. I've merged any recent changes into `contributing.rst` and removed `testing.rst` now.

jneeven reviewed (approved):
LGTM! @ttuffin thanks for taking care of this

unode commented on AUTHORS.rst:
```diff
@@ -2,20 +2,18 @@
-* tgly307 <tgly307@163.com>
+* Alex Tzonkov <alex.tzonkov@gmail.com>
+* Jelmer Neeven <https://github.com/jneeven>
+* Renato Alves <https://github.com/unode>
```
```suggestion
* Renato Alves <alves.rjc@gmail.com>
```

attzonko commented:
@unode Should we wait for the Migration doc or merge this and open a new one specifically for migration?

unode commented:
Please go ahead and merge. I don't think we need to wait on this for the following steps.
